contracts/token/ERC721/ERC721.sol

```diff
@@ -112,28 +112,28 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
     /**
      * @dev See {IERC721-ownerOf}.
      */
-    function ownerOf(uint256 tokenId) public view override returns (address) {
+    function ownerOf(uint256 tokenId) public view virtual override returns (address) {
```

Comment:
Not sure how I feel about this one.

Reply:
So,

This function reverts when to token doesn't exist (not sure if that is part of the protocole). So the usecase that returns address(0) if expirect actually make sens: "the token exist, but isn't owned by anyone". Still there would be an issue if someone was to transfer an expired token, since that would trigger events (and hook) with `from = address(0)`, and it would try to remove the tokenid from the `_holderTokens[0]` set.

So if this is made virtual, it should not be used as a getter when doing `_transfer` or `_burn` (nor in _approve, which would technically still be triggered if other function are overloaded to give approval to tokens owner by 0).

This is a tricky one. Not as obviously breaking as balanceOf, but still sensitive. The thing is that, unlike balanceOf (and provided we don't use this getter in the place described before), a user could overload that in a way that would keep the registry "mostly" standard without having to overload other functions.

If we have it virtual, we should have a big disclaimer. If we have it non-virtual, we need a clear argument as to why not (this was requested)

Reply:
>So the usecase that returns address(0) if expired

Note that the [example](https://github.com/ensdomains/ethregistrar/blob/75b4515b33ff06f4e5e980c36a9a5214e7d1bd4c/contracts/BaseRegistrarImplementation.sol#L46-L49) the user shared was:

```solidity
    function ownerOf(uint256 tokenId) public view returns (address) {
        require(expiries[tokenId] > now);
        return super.ownerOf(tokenId);
    }
```

So it actually reverts if expired. I don't really think reverting here means anything for the purposes of ERC721 semantics. The token may have an owner but the contract might still allow someone else to reclaim it if it has expired. So if I were to design the contract I wouldn't worry about `ownerOf`... But in any case, it's true that we could internally use a private `_ownerOf` function (or read directly from storage), and allow people if they want to override the public function. I don't see an issue with that.

We could even do the same thing for `ERC20.balanceOf`... People can override the public function if they want but we internally keep reading from and writing to storage. If they want to change those semantics they can override `_transfer` or the before hook.